csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/InferenceTest.netcore.cs

```diff
@@ -67,6 +74,196 @@ public void CanCreateAndDisposeSessionWithModelPath()
+        {
+            var model = TestDataLoader.LoadModelFromEmbeddedResource("squeezenet.onnx");
+
+            using (var cleanUp = new DisposableListTest<IDisposable>())
+            {
+                // Set the graph optimization level for this session.
+                SessionOptions options = new SessionOptions();
+                cleanUp.Add(options);
+                options.GraphOptimizationLevel = graphOptimizationLevel;
+                if (enableParallelExecution) options.ExecutionMode = ExecutionMode.ORT_PARALLEL;
+
+                var session = new InferenceSession(model, options);
+                cleanUp.Add(session);
+
+                using var runOptions = new RunOptions();
+                using var inputOrtValues = new DisposableListTest<DisposableTestPair<OrtValue>>(session.InputMetadata.Count);
+                var inputMeta = session.InputMetadata;
+                var outputMeta = session.OutputMetadata;
+
+                float[] expectedOutput = TestDataLoader.LoadTensorFromEmbeddedResource("bench.expected_out");
+                long[] expectedDimensions = { 1, 1000, 1, 1 };  // hardcoded for now for the test data
+                ReadOnlySpan<long> expectedOutputDimensions = expectedDimensions;
+                string[] expectedOutputNames = new string[] { "softmaxout_1" };
+
+                float[] inputData = TestDataLoader.LoadTensorFromEmbeddedResource("bench.in"); // this is the data for only one input tensor for this model
+
+                foreach (var name in inputMeta.Keys)
+                {
+                    Assert.Equal(typeof(float), inputMeta[name].ElementType);
+                    Assert.True(inputMeta[name].IsTensor);
+                    var tensor = DotnetTensors.Tensor.Create<float>(inputData, inputMeta[name].Dimensions.Select(x => (nint)x).ToArray());
+                    inputOrtValues.Add(new DisposableTestPair<OrtValue>(name, OrtValue.CreateTensorValueFromSystemNumericsTensorObject<float>(tensor)));
+
+                }
+
+                runOptions.LogId = "CsharpTest";
+                runOptions.Terminate = false;  // TODO: Test terminate = true, it currently crashes
+                runOptions.LogSeverityLevel = OrtLoggingLevel.ORT_LOGGING_LEVEL_ERROR;
+                // Run inference with named inputs and outputs created with in Run()
+                using (var results = session.Run(runOptions, inputOrtValues.Select(x => x.Key).ToList(), inputOrtValues.Select(x => x.Value).ToList(), new List<string>(["softmaxout_1"])))  // results is an IDisposableReadOnlyCollection<OrtValue> container
+                {
+                    // validate the results
+                    foreach (var r in results)
+                    {
+                        Assert.Single(results);
+
+                        ValidateRunResult(r, expectedOutput, expectedDimensions);
+                    }
+                }
+
+                session.Dispose();
```

Comment:
>session [](http://example.com/codeflow?start=16&length=7)

session may be disposed twice as it is in the cleanup list.
Also an explicit call to Dispose() is usually an indication of exception safety problem and a memory leak.

Reply:
Fixed.